include/veriblock/blockchain/pop/pop_state_machine.hpp

```diff
@@ -26,19 +26,34 @@ struct PopStateMachine {
 
   PopStateMachine(ProtectedTree& ed,
                   ProtectingBlockTree& ing,
+                  PopStorage& storage,
                   height_t startHeight = 0)
-      : ed_(ed), ing_(ing), startHeight_(startHeight) {}
+      : ed_(ed), ing_(ing), storage_(storage), startHeight_(startHeight) {}
 
   bool applyBlock(index_t& index, ValidationState& state) {
     std::vector<CommandPtr> executed;
-    for (auto& cg : index.commands) {
+    for (const auto& pid : index.payloadIds) {
+      payloads_t payloads;
+      bool ret = storage_.loadPayloads(pid, payloads);
+      if (!ret) {
+        //TODO: error message
+        continue;
+      }
+      std::vector<CommandPtr> commands{};
+      ed_.payloadsToCommands(payloads, commands);
       VBK_LOG_DEBUG("Applying payload %s from block %s",
-                    cg.id.toHex(),
+                    pid.toHex(),
                     index.toShortPrettyString());
-      for (auto& cmd : cg) {
+
+      for (auto& cmd : commands) {
         if (!cmd->Execute(state)) {
           // invalidate command group
-          cg.valid = false;
+          payloads.valid = false;
+          ret = storage_.savePayloads(payloads);
```

Comment:
you do not have to update storage here. we only update storage periodically from `save` function

Reply:
How do we add valid flag then?